### ISSUE TITLE
Hook callbacks are called random number of times

### ISSUE THREAD

**Issue by dominioon** [orientechnologies/orientdb#6798]
## OrientDB Version, operating system, or hardware.
-  [x] v2.2.11

## Operating System
- [X] Linux

## Expected behavior and actual behavior
The callbacks like `onRecordAfterCreate()` should be called only once for every insert.

## Steps to reproduce the problem
```
public class MyHook extends ODocumentHookAbstract {
  @Override
  public DISTRIBUTED_EXECUTION_MODE getDistributedExecutionMode() {
    return DISTRIBUTED_EXECUTION_MODE.TARGET_NODE;
  }
  @Override
  public void onRecordAfterCreate(ODocument iDocument) {
    System.out.println("Created " + iDocument.toJSON());
  }
}
```
```
  for (int i = 0; i < 10; i++) {
    try (ODatabaseDocumentTx db = dbPool.acquire()) {
      db.save(new ODocument("SomeTable").field("entry", i));
    }
  }
```

Result:
```
Created {"@type":"d","@rid":"#17:0","@version":1,"@class":"SomeTable","entry":0}
Created {"@type":"d","@rid":"#18:0","@version":1,"@class":"SomeTable","entry":1}
Created {"@type":"d","@rid":"#19:0","@version":1,"@class":"SomeTable","entry":2}
Created {"@type":"d","@rid":"#19:0","@version":1,"@class":"SomeTable","entry":2}
Created {"@type":"d","@rid":"#20:0","@version":1,"@class":"SomeTable","entry":3}
Created {"@type":"d","@rid":"#20:0","@version":1,"@class":"SomeTable","entry":3}
Created {"@type":"d","@rid":"#17:1","@version":1,"@class":"SomeTable","entry":4}
Created {"@type":"d","@rid":"#17:1","@version":1,"@class":"SomeTable","entry":4}
Created {"@type":"d","@rid":"#17:1","@version":1,"@class":"SomeTable","entry":4}
Created {"@type":"d","@rid":"#18:1","@version":1,"@class":"SomeTable","entry":5}
Created {"@type":"d","@rid":"#18:1","@version":1,"@class":"SomeTable","entry":5}
Created {"@type":"d","@rid":"#18:1","@version":1,"@class":"SomeTable","entry":5}
Created {"@type":"d","@rid":"#19:1","@version":1,"@class":"SomeTable","entry":6}
Created {"@type":"d","@rid":"#19:1","@version":1,"@class":"SomeTable","entry":6}
Created {"@type":"d","@rid":"#19:1","@version":1,"@class":"SomeTable","entry":6}
Created {"@type":"d","@rid":"#19:1","@version":1,"@class":"SomeTable","entry":6}
Created {"@type":"d","@rid":"#20:1","@version":1,"@class":"SomeTable","entry":7}
Created {"@type":"d","@rid":"#20:1","@version":1,"@class":"SomeTable","entry":7}
Created {"@type":"d","@rid":"#20:1","@version":1,"@class":"SomeTable","entry":7}
Created {"@type":"d","@rid":"#20:1","@version":1,"@class":"SomeTable","entry":7}
Created {"@type":"d","@rid":"#17:2","@version":1,"@class":"SomeTable","entry":8}
Created {"@type":"d","@rid":"#17:2","@version":1,"@class":"SomeTable","entry":8}
Created {"@type":"d","@rid":"#17:2","@version":1,"@class":"SomeTable","entry":8}
Created {"@type":"d","@rid":"#17:2","@version":1,"@class":"SomeTable","entry":8}
Created {"@type":"d","@rid":"#17:2","@version":1,"@class":"SomeTable","entry":8}
Created {"@type":"d","@rid":"#18:2","@version":1,"@class":"SomeTable","entry":9}
Created {"@type":"d","@rid":"#18:2","@version":1,"@class":"SomeTable","entry":9}
Created {"@type":"d","@rid":"#18:2","@version":1,"@class":"SomeTable","entry":9}
Created {"@type":"d","@rid":"#18:2","@version":1,"@class":"SomeTable","entry":9}
Created {"@type":"d","@rid":"#18:2","@version":1,"@class":"SomeTable","entry":9}
```

**Comment by tglman** [comment]
Hi @dominioon,

how do you register the hook in the database instance ?

here is a example test case:
```
package test;

import org.junit.After;
import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;

import com.orientechnologies.orient.core.db.OPartitionedDatabasePool;
import com.orientechnologies.orient.core.db.document.ODatabaseDocument;
import com.orientechnologies.orient.core.db.document.ODatabaseDocumentTx;
import com.orientechnologies.orient.core.hook.ODocumentHookAbstract;
import com.orientechnologies.orient.core.hook.ORecordHook.DISTRIBUTED_EXECUTION_MODE;
import com.orientechnologies.orient.core.record.impl.ODocument;

public class HookTest {

  public class MyHook extends ODocumentHookAbstract {
    @Override
    public DISTRIBUTED_EXECUTION_MODE getDistributedExecutionMode() {
      return DISTRIBUTED_EXECUTION_MODE.TARGET_NODE;
    }

    @Override
    public void onRecordAfterCreate(ODocument iDocument) {
      System.out.println("Created " + iDocument.toJSON());
      count++;
    }
  }

  private int count = 0;

  @Before
  public void before() {
    ODatabaseDocument db = new ODatabaseDocumentTx("memory:test");
    db.create();
    db.getMetadata().getSchema().createClass("Test");
    db.close();
  }

  @After
  public void after() {
    ODatabaseDocument db = new ODatabaseDocumentTx("memory:test");
    db.open("admin", "admin");
    db.drop();
  }

  @Test
  public void test() {
    OPartitionedDatabasePool pool = new OPartitionedDatabasePool("memory:test", "admin", "admin");
    for (int i = 0; i < 10; i++) {
      try (ODatabaseDocument some = pool.acquire()) {
        MyHook aa = new MyHook();
        some.registerHook(aa);
        some.save(new ODocument("Test").field("entry", i));
        //If you comment out this you get your problem ;)
        some.unregisterHook(aa);
      }
    }
    Assert.assertEquals(10, count);
    pool.close();

  }

}
```

**Comment by dominioon** [comment]
Hi, @tglman,
The hook was registered via orientdb-server-config.xml 
    <hooks>
        <hook class="com.MyHook" position="REGULAR"/>
    </hooks>

**Comment by tglman** [comment]
hi @dominioon,

so you are registering the hook server side, and the log is server side, are you running operations on remote or embedding the server?

Bye

**Comment by dominioon** [comment]
Hi @tglman, 

the server is embedded. See the example app from https://github.com/orientechnologies/orientdb/issues/6718.  Sorry, Im not near computer and cant post the full example atm by myself.

**Comment by tglman** [comment]
hi @davidmoten,

I reproduced this writing a test with an embedded server, ad I fixed it also, is going to be released in 2.2.12.

Thank you for reporting this.
Regards

**Comment by tglman** [comment]
Hi,

The 2.2.12 has been released with this fix, closing

Bye
